File: packages/shared/src/schemas/prompt.schemas.ts
import { z } from '@hono/zod-openapi';
// --- Base Schema for a Prompt ---
// Represents the structure of a single Prompt object as returned by the API
export const PromptSchema = z.object({
    id: z.string().min(1).openapi({ example: 'p1a2b3c4-e5f6-7890-1234-567890abcdef', description: 'Prompt ID' }),
    name: z.string().openapi({ example: 'Code Refactoring Prompt', description: 'Prompt name' }),
    content: z.string().openapi({ example: 'Refactor the following code to be more efficient: {code}', description: 'Prompt content template' }),
    projectId: z.string().min(1).optional().openapi({ example: 'project-123', description: 'Optional Project ID this prompt is linked to (contextual)' }),
    createdAt: z.string().datetime().openapi({ example: '2024-02-01T10:00:00.000Z', description: 'Creation timestamp (ISO 8601)' }),
    updatedAt: z.string().datetime().openapi({ example: '2024-02-01T10:05:00.000Z', description: 'Last update timestamp (ISO 8601)' })
}).openapi('Prompt');

// --- Request Body Schemas ---
export const CreatePromptBodySchema = z.object({
    // Allow projectId to be optional during creation, linking can happen separately or via this field
    projectId: z.string().min(1).optional().openapi({ example: 'project-456', description: 'Optional Project ID to link the prompt to upon creation' }),
    name: z.string().min(1).openapi({ example: 'My New Prompt' }),
    content: z.string().min(1).openapi({ example: 'Translate this text: {text}' }),
}).openapi('CreatePromptRequestBody');

export const UpdatePromptBodySchema = z.object({
    name: z.string().min(1).optional().openapi({ example: 'Updated Prompt Name' }),
    content: z.string().min(1).optional().openapi({ example: 'Updated content: {variable}' }),
}).refine(data => data.name || data.content, {
    message: "At least one of name or content must be provided for update"
}).openapi('UpdatePromptRequestBody');

// --- Request Parameter Schemas ---
export const PromptIdParamsSchema = z.object({
    promptId: z.string().min(1).openapi({
        param: { name: 'promptId', in: 'path' },
        example: 'p1a2b3c4-e5f6-7890-1234-567890abcdef',
        description: 'The UUID of the prompt'
    })
}).openapi('PromptIdParams');

export const ProjectIdParamsSchema = z.object({
    projectId: z.string().min(1).openapi({
        param: { name: 'projectId', in: 'path' },
        example: 'project-789',
        description: 'The ID of the project'
    })
}).openapi('ProjectIdParams');

export const ProjectAndPromptIdParamsSchema = z.object({
    projectId: z.string().min(1).openapi({
        param: { name: 'projectId', in: 'path' },
        example: 'project-abc',
        description: 'The ID of the project'
    }),
    promptId: z.string().min(1).openapi({
        param: { name: 'promptId', in: 'path' },
        example: 'p1a2b3c4-e5f6-7890-1234-567890abcdef',
        description: 'The ID of the prompt'
    })
}).openapi('ProjectAndPromptIdParams');


// --- Response Schemas ---
export const PromptResponseSchema = z.object({
    success: z.literal(true),
    data: PromptSchema // Use the base Prompt schema
}).openapi('PromptResponse');

export const PromptListResponseSchema = z.object({
    success: z.literal(true),
    data: z.array(PromptSchema) // Array of Prompt schemas
}).openapi('PromptListResponse');


// --- Keep the original structure for potential direct use (optional) ---
// This might be redundant now with the OpenAPI-enhanced schemas above
export const promptApiValidation = {
    create: {
        body: CreatePromptBodySchema // Use the OpenAPI schema
    },
    list: {
        params: ProjectIdParamsSchema // Use the OpenAPI schema
    },
    getOrDelete: {
        params: PromptIdParamsSchema // Use the OpenAPI schema
    },
    update: {
        params: PromptIdParamsSchema, // Use the OpenAPI schema
        body: UpdatePromptBodySchema // Use the OpenAPI schema
    },
    // Specific schemas for project/prompt linking
    addOrRemoveFromProject: {
        params: ProjectAndPromptIdParamsSchema // Use the combined schema
    }
} as const;

// Export types if needed elsewhere
export type CreatePromptBody = z.infer<typeof CreatePromptBodySchema>;
export type UpdatePromptBody = z.infer<typeof UpdatePromptBodySchema>;
export type PromptIdParams = z.infer<typeof PromptIdParamsSchema>;
export type ProjectIdParams = z.infer<typeof ProjectIdParamsSchema>;
export type ProjectAndPromptIdParams = z.infer<typeof ProjectAndPromptIdParamsSchema>;


// --- Request Body Schema ---
export const OptimizePromptRequestSchema = z.object({
    userContext: z.string().min(1).openapi({
        example: "Make my login form better.",
        description: "The user's initial prompt or context to be optimized."
    }),
    // Add other potential fields if needed, e.g., targetAudience, desiredTone etc.
}).openapi('OptimizePromptRequest');


// --- Response Schema ---
export const OptimizePromptResponseSchema = z.object({
    success: z.literal(true).openapi({ description: 'Indicates successful optimization' }),
    data: z.object({
        optimizedPrompt: z.string().openapi({
            example: "Optimize the user experience for the login form, focusing on clarity, security, and accessibility. Suggest improvements for field labels, error handling, password requirements display, and button text.",
            description: "The optimized prompt generated by the service."
        })
    })
}).openapi('OptimizePromptResponse');

// --- Validation Object (optional, follows pattern) ---
export const promptimizerApiValidation = {
    optimize: {
        body: OptimizePromptRequestSchema
    }
} as const;

export const PromptProjectSchema = z.object({
    id: z.string(),
    promptId: z.string(),
    projectId: z.string(),
});


// Export types if needed elsewhere
export type OptimizePromptRequest = z.infer<typeof OptimizePromptRequestSchema>;
export type Prompt = z.infer<typeof PromptSchema>;
export type PromptList = z.infer<typeof PromptListResponseSchema>;
export type PromptResponse = z.infer<typeof PromptResponseSchema>;
export type PromptProject = z.infer<typeof PromptProjectSchema>;

File: packages/python_backend/app/schemas/prompt_schemas.py
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field, ConfigDict, field_validator, model_validator
from enum import Enum
from app.utils.storage_timestap_utils import convert_timestamp_to_ms_int, convert_id_to_int, convert_optional_id_to_int

# --- Prompt Schemas ---
# Last 5 changes:
# 1. Initial conversion from Zod to Pydantic.
# 2. Handled .datetime(), .optional(), .refine() (with model_validator).
# 3. Mapped .openapi() metadata including param details.
# 4. Created Pydantic models for all request/response bodies and params.
# 5. Imported ProjectIdParams placeholder.
# 6. Changed ID and timestamp fields to int (Unix ms) and added validators.

# Placeholder for ProjectIdParamsSchema if it were in a different file
# from .project_schemas import ProjectIdParamsSchema # Assuming this path
class ProjectIdParams(BaseModel):
    project_id: int = Field(..., validation_alias="projectId", serialization_alias="projectId", json_schema_extra={"param": {"name": "projectId", "in": "path"}}, example=1677657600000, description="The ID of the project (Unix ms)")
    model_config = ConfigDict(populate_by_name=True)
    _validate_project_id = field_validator('project_id', mode='before')(convert_id_to_int)

class Prompt(BaseModel):
    id: int = Field(..., example=1675248000000, description="Prompt ID (Unix ms)")
    name: str = Field(..., example="Code Refactoring Prompt", description="Prompt name")
    content: str = Field(..., example="Refactor the following code to be more efficient: {code}", description="Prompt content template")
    project_id: Optional[int] = Field(None, validation_alias="projectId", serialization_alias="projectId", example=1677657600000, description="Optional Project ID this prompt is linked to (Unix ms)")
    created: int = Field(..., validation_alias="created", serialization_alias="created", example=1675248000000, description="Creation timestamp (Unix ms)")
    updated: int = Field(..., validation_alias="updated", serialization_alias="updated", example=1675248300000, description="Last update timestamp (Unix ms)")
    model_config = ConfigDict(title="Prompt", populate_by_name=True)

    _validate_id = field_validator('id', mode='before')(convert_id_to_int)
    _validate_project_id = field_validator('project_id', mode='before')(convert_optional_id_to_int)
    _validate_timestamps = field_validator('created', 'updated', mode='before')(convert_timestamp_to_ms_int)

class CreatePromptBody(BaseModel):
    project_id: Optional[int] = Field(None, validation_alias="projectId", serialization_alias="projectId", example=1677657600000, description="Optional Project ID to link the prompt to upon creation (Unix ms)")
    name: str = Field(..., min_length=1, example="My New Prompt")
    content: str = Field(..., min_length=1, example="Translate this text: {text}")
    model_config = ConfigDict(title="CreatePromptRequestBody", populate_by_name=True)

    _validate_project_id = field_validator('project_id', mode='before')(convert_optional_id_to_int)

class UpdatePromptBody(BaseModel):
    name: Optional[str] = Field(None, min_length=1, example="Updated Prompt Name")
    content: Optional[str] = Field(None, min_length=1, example="Updated content: {variable}")

    @model_validator(mode='after')
    def check_at_least_one_field(self):
        if not self.name and not self.content:
            raise ValueError("At least one of name or content must be provided for update")
        return self
    model_config = ConfigDict(title="UpdatePromptRequestBody")

class PromptIdParams(BaseModel):
    prompt_id: int = Field(..., validation_alias="promptId", serialization_alias="promptId", json_schema_extra={"param": {"name": "promptId", "in": "path"}}, example=1675248000000, description="The ID of the prompt (Unix ms)")
    model_config = ConfigDict(title="PromptIdParams", populate_by_name=True)
    _validate_prompt_id = field_validator('prompt_id', mode='before')(convert_id_to_int)

class ProjectAndPromptIdParams(BaseModel):
    project_id: int = Field(..., validation_alias="projectId", serialization_alias="projectId", json_schema_extra={"param": {"name": "projectId", "in": "path"}}, example=1677657600000, description="The ID of the project (Unix ms)")
    prompt_id: int = Field(..., validation_alias="promptId", serialization_alias="promptId", json_schema_extra={"param": {"name": "promptId", "in": "path"}}, example=1675248000000, description="The ID of the prompt (Unix ms)")
    model_config = ConfigDict(title="ProjectAndPromptIdParams", populate_by_name=True)
    _validate_ids = field_validator('project_id', 'prompt_id', mode='before')(convert_id_to_int)

class PromptResponse(BaseModel):
    success: Literal[True] = True
    data: Prompt
    model_config = ConfigDict(title="PromptResponse")

class PromptListResponse(BaseModel):
    success: Literal[True] = True
    data: List[Prompt]
    model_config = ConfigDict(title="PromptListResponse")

class OptimizeUserInputRequest(BaseModel):
    project_id: int = Field(..., validation_alias="projectId", serialization_alias="projectId", example=1677657600000, description="The ID of the project (Unix ms)")
    user_context: str = Field(..., min_length=1, validation_alias="userContext", serialization_alias="userContext", example="Make my login form better.", description="The user's initial prompt or context to be optimized.")
    model_config = ConfigDict(title="OptimizePromptRequest", populate_by_name=True)
    _validate_project_id = field_validator('project_id', mode='before')(convert_id_to_int)

class OptimizedPromptData(BaseModel):
    optimized_prompt: str = Field(..., validation_alias="optimizedPrompt", serialization_alias="optimizedPrompt", example="Optimize the user experience for the login form, focusing on clarity, security, and accessibility. Suggest improvements for field labels, error handling, password requirements display, and button text.", description="The optimized prompt generated by the service.")
    model_config = ConfigDict(populate_by_name=True)

class OptimizePromptResponse(BaseModel):
    success: Literal[True] = Field(..., description="Indicates successful optimization")
    data: OptimizedPromptData
    model_config = ConfigDict(title="OptimizePromptResponse")

class PromptProject(BaseModel):
    id: int
    prompt_id: int = Field(..., validation_alias="promptId", serialization_alias="promptId")
    project_id: int = Field(..., validation_alias="projectId", serialization_alias="projectId")
    created: Optional[int] = Field(None, validation_alias="created", serialization_alias="created")
    model_config = ConfigDict(populate_by_name=True)

    _validate_ids = field_validator('id', 'prompt_id', 'project_id', mode='before')(convert_id_to_int)
    _validate_created_at = field_validator('created', mode='before')(convert_timestamp_to_ms_int)
